### DISCUSSION TITLE
pragmatically change x-range

### DISCUSSION THREAD

**Discussion by baum-sparton** [patrykandpatrick/vico#929] [Questions] [answered]
### Question

Thank you for all the hard work and responsive support you provide for Vico.  I might be missing something here.  I am looking to dynamically change the x-range based on user input.  Imagine a situation where certain ranges have particular meaning, -10<x<10, -50<x<50, etc, and there is some UI control that allows the user to select their desired range.

I'm using the following to set the range of the x-axis.

    rangeProvider = CartesianLayerRangeProvider.fixed(minX = minX, maxX = maxX)

from within rememberLineCartesianLayer, where minX and maxX are the values from the user interaction.  This works as expected when we first render the screen.  However, regardless of if minX and maxX are local variables or mutable states, changing them does not update the graph.  I have found that in order to get the graph to reflect the changes I need to run modelProducer.runTransaction again to cause a change in the layers.  I believe this is all expected based on the implementation.  I would however prefer not doing that to avoid overhead preparing the data which occurs during my current implementation of modelProducer.runTransaction.  Is there another way to change these values and trigger a redraw?

Note: I originally came across the issue because I'm calling modelProducer.runTransaction from with a LaunchedEffect that get re-ran when my data changes:

```
LaunchedEffect(graphData) {
    withContext(Dispatchers.Default) {
        modelProducer.runTransaction { ... }}}
```

As such as new data was coming in the graph was updating, but as the user selected the x-range it wasn't.  Thanks 

### Vico version(s)

2.0.0-beta.1

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @baum-sparton. Running a `Transaction` is the only correct way to change the ranges. However, it's not simply about running **some** `Transaction` to cause the changes to be reflected. (You don't need just a redraw, which you can achieve with state, but rather a `CartesianChartRanges` update.) That would be more of a workaround, not to mention that it wouldn't work without a data update, as `Transaction`s that don't change anything are ignored. Also, it wouldn't provide sufficient synchronization.

Rather, you should use extras (which are specifically intended for such purposes), have a single `CartesianLayerRangeProvider` that reads them, and update the ranges by running `Transaction`s that change the extras' values. `CartesianLayerRangeProvider.fixed` is not suitable here.

This is covered [in the wiki](https://patrykandpatrick.com/vico/wiki/cartesian-charts/layers/#cartesianlayerrangeprovider). You can find a broader overview that also discusses the design of the API [here](https://github.com/patrykandpatrick/vico/discussions/674#discussioncomment-9444387).

> I would however prefer not doing that to avoid overhead preparing the data which occurs during my current implementation of modelProducer.runTransaction.

You should update your code to avoid this. The data should be transformed only when it changes. This way, you'll be able to run `Transaction`s that update the ranges without unnecessarily retransforming the data and introducing unwanted overhead.

**Reply by baum-sparton** [reply]
In case it helps others, here is what I put together for a CartesianLayerRangeProvider.

I defined the following extra keys:
```
    val minXExtraKey = remember { ExtraStore.Key<Double>() } 
    val maxXExtraKey = remember { ExtraStore.Key<Double>() }
    val xStepExtraKey = remember { ExtraStore.Key<Double>() }
```

I then call out a custom getXStep and rangeProvider when defining the graph.

`rangeProvider = remember { getCustomRangeProvider(minXExtraKey, maxXExtraKey) }`

```
getXStep = { it ->
     val step = it.extraStore.getOrNull(xStepExtraKey)
     step?: 1.0
```

With the range provider I make use of NaN to indicate the app should choose the range (since I couldn't have a Double? extra).  Here is the custom range provider.

```
private fun getCustomRangeProvider(
    minXExtraKey: ExtraStore.Key<Double>,
    maxXExtraKey: ExtraStore.Key<Double>
): CartesianLayerRangeProvider {
    return object : CartesianLayerRangeProvider {
        override fun getMinX(minX: Double, maxX: Double, extraStore: ExtraStore): Double {
            val min = extraStore.getOrNull(minXExtraKey)
            if (min == null) {
                Log.e(TAG, "Missing minX extra.")
                return super.getMinX(minX, maxX, extraStore)
            } else if (min.isNaN()) {
                return super.getMinX(minX, maxX, extraStore)
            } else {
                return min
            }
        }

        override fun getMaxX(minX: Double, maxX: Double, extraStore: ExtraStore): Double {
            val max = extraStore.getOrNull(maxXExtraKey)
            if (max == null) {
                Log.e(TAG, "Missing maxX extra.")
                return super.getMaxX(minX, maxX, extraStore)
            } else if (max.isNaN()) {
                return super.getMaxX(minX, maxX, extraStore)
            } else {
                return max
            }
        }
    }
}
```

While I need to do some more testing, this does seem to produce the desired behavior.

**Reply by baum-sparton** [reply]
I do seem to be experiencing an occasional 'dropped' transaction.  What I mean by this is that `modelProducer.runTransaction` is called, but the graph does not refresh.  I'm observing the calls by logging placed within runTransation and the getMinX of my custom CartesianLayerRangeProvider.  I've seen the behavior occur with the transaction changing the series data or the extra data.  The result is either a blank (white rectangle) or a stale graph.

The behavior currently seems sporadic as I haven't yet been able to nail down a sequence of events in which this occurs.  Are there known situations where transaction changes don't result in refreshing the UI?  Does the library provide a way to peak at or log its behavior.

I'm suspecting the issues is occurring when screens are first displayed and recompositions are occurring due to flow backed states changing from the initial values to the values provided by the flow.  The series data, axes ranges, and units are provided this way through flows from the view model.  As these individual state values are loaded, the transactions are triggered, sometimes 3 or 4 in a row.

**Reply by Gowsky** [reply]
> In case it helps others, here is what I put together for a CartesianLayerRangeProvider.

Generally, this should work, but you seem to be using the API somewhat incorrectly. As a result, the code is longer and less direct than it should be. Solutions like the `NaN` workaround aren't necessary. Please show us the logic that runs the `Transaction`s (including adding the extras) so that we have the full picture and can describe the recommended corrections efficiently.

> I do seem to be experiencing an occasional 'dropped' transaction.

It sounds like there's an issue with `Transaction` handling on your end. Seeing the updated `Transaction` logic will help us address this too.

**Reply by baum-sparton** [reply]
The `NaN `is just mechanism I'm using in the custom `CartesianLayerRangeProvider `to signify that the user wants the graph to choose the best range.

Here is a larger snippet of the code which exhibits the issue and includes the transaction.  There is one launched effect which prepares the graph data when the input data gets modified.  Then there is a second launched effect which does the transaction when the prepared data has changed or a parameter controlling the x-axis has changed.

Thanks for the help and sorry for the amount of code.

```
private data class SeriesData(
    val x: List<Double>,
    val y: List<Double>,
)

private data class GraphState(
    val series: List<SeriesData>
) {
    constructor() : this(listOf(SeriesData(listOf(0.0), listOf(0.0))))
}

@Composable
internal fun ScatterPlot(
    modifier: Modifier,
    events: List<Event>,
    calibrate: Boolean,
    timeWindow: Long
) {
    val myPoint = remember { getColoredPoint(Color.Red) }

    val minXExtraKey = remember { ExtraStore.Key<Double>() }
    val maxXExtraKey = remember { ExtraStore.Key<Double>() }
    val xStepExtraKey = remember { ExtraStore.Key<Double>() }
    val modelProducer = remember { CartesianChartModelProducer() }

    var graphState by remember { mutableStateOf<GraphState>(GraphState()) }

    LaunchedEffect(events, calibrate) {
        withContext(Dispatchers.Default) {
            graphState = processData(events, calibrate)
        }
    }

    LaunchedEffect(graphState, timeWindow) {
        withContext(Dispatchers.Default) {
            modelProducer.runTransaction {
                lineSeries {
                    for (s in graphState.series) {
                        series(s.x, s.y)
                    }
                }

                var minX: Double = Double.NaN
                var maxX: Double = Double.NaN
                var xStep: Double = 1.0
                if (timeWindow != 0L) {
                    minX = -timeWindow.toDouble()
                    maxX = 0.0
                    xStep = (timeWindow/4).toDouble()
                }

                extras { extraStore ->
                    extraStore[minXExtraKey] = minX
                    extraStore[maxXExtraKey] = maxX
                    extraStore[xStepExtraKey] = xStep
                }
            }
        }
    }

    CartesianChartHost(
        chart = rememberCartesianChart(
            rememberLineCartesianLayer(
                LineCartesianLayer.LineProvider.series(
                    LineCartesianLayer.rememberLine(
                        pointProvider = remember {
                            object : LineCartesianLayer.PointProvider {
                                override fun getPoint(
                                    entry: LineCartesianLayerModel.Entry,
                                    seriesIndex: Int,
                                    extraStore: ExtraStore,
                                ): LineCartesianLayer.Point {
                                    return myPoint
                                }

                                override fun getLargestPoint(extraStore: ExtraStore): LineCartesianLayer.Point =
                                    myPoint
                            }
                        },
                        fill = LineCartesianLayer.LineFill.single(fill = fill(Color.Transparent)),
                        thickness = 0.dp,
                        areaFill = null
                    )
                ),
                rangeProvider = remember { getCustomRangeProvider(minXExtraKey, maxXExtraKey) }
            ),
            getXStep = { it ->
                val step = it.extraStore.getOrNull(xStepExtraKey)
                step?: 1.0
            },
            startAxis = VerticalAxis.rememberStart(
                titleComponent = rememberTextComponent(),
                title= "y title",
                itemPlacer = remember { VerticalAxis.ItemPlacer.count(count = { 4 }) }
            ),
            bottomAxis = HorizontalAxis.rememberBottom(
                titleComponent = rememberTextComponent(),
                title= "x title",
                guideline = null,
                itemPlacer = HorizontalAxis.ItemPlacer.aligned()
            ),
            layerPadding = cartesianLayerPadding(scalableStartPadding = 8.dp, scalableEndPadding = 8.dp),
        ),
        modelProducer = modelProducer,
        modifier = modifier,
        zoomState = rememberVicoZoomState(
            zoomEnabled = true,
            initialZoom = Zoom.Content
        ),
    )
}

private fun getColoredPoint(color: Color) : LineCartesianLayer.Point {
    return LineCartesianLayer.point(
        component = shapeComponent(
            shape = CorneredShape.Pill,
            color = color,
        ),
        size = 4.dp,
    )
}

private fun getCustomRangeProvider(
    minXExtraKey: ExtraStore.Key<Double>,
    maxXExtraKey: ExtraStore.Key<Double>
): CartesianLayerRangeProvider {
    return object : CartesianLayerRangeProvider {
        override fun getMinX(minX: Double, maxX: Double, extraStore: ExtraStore): Double {
            val min = extraStore.getOrNull(minXExtraKey)
            if (min == null) {
                Log.e(TAG, "Missing minX extra.")
                return super.getMinX(minX, maxX, extraStore)
            } else if (min.isNaN()) {
                return super.getMinX(minX, maxX, extraStore)
            } else {
                return min
            }
        }

        override fun getMaxX(minX: Double, maxX: Double, extraStore: ExtraStore): Double {
            val max = extraStore.getOrNull(maxXExtraKey)
            if (max == null) {
                Log.e(TAG, "Missing maxX extra.")
                return super.getMaxX(minX, maxX, extraStore)
            } else if (max.isNaN()) {
                return super.getMaxX(minX, maxX, extraStore)
            } else {
                return max
            }
        }
    }
}
```

**Reply by Gowsky** [reply, marked as answer]
> The `NaN` is just mechanism I'm using in the custom `CartesianLayerRangeProvider` to signify that the user wants the graph to choose the best range.

Understood, but it's not the optimal approach.

First of all, if you always add an extra, it's always there, so there's no need to use `getOrNull`. Therefore, in `getXStep`, you should simply use this:

```kt
it.extraStore[xStepExtraKey]
```

For the ranges, instead of always adding the extras, using `getOrNull` unnecessarily, and using `NaN` to indicate when the defaults should be used, you should add the extras conditionally and use `getOrNull` to handle cases when the optional extras aren't present. This is the purpose of `getOrNull`.

```kt
var minX: Double? = null
var maxX: Double? = null
var xStep: Double = 1.0

...

extras { extraStore ->
    minX?.let { minX -> extraStore[minXExtraKey] = minX }
    maxX?.let { maxX -> extraStore[maxXExtraKey] = maxX }
    extraStore[xStepExtraKey] = xStep
}
```

```kt
override fun getMinX(minX: Double, maxX: Double, extraStore: ExtraStore) =
    extraStore.getOrNull(minXExtraKey) ?: super.getMinX(minX, maxX, extraStore)

override fun getMaxX(minX: Double, maxX: Double, extraStore: ExtraStore) =
    extraStore.getOrNull(maxXExtraKey) ?: super.getMaxX(minX, maxX, extraStore)
```

Such an approach is simpler and more direct.

This should make it clear why null extras aren't allowed. If there's no value, you simply don't add the extra.

Moving on to the dropped `Transaction`:

First of all, if all the data is in the `ViewModel`, then that's where you should store the `CartesianChartModelProducer` and run `Transaction`s. `ScatterPlot` would then have just one argument (`modelProducer`). Storing the `CartesianChartModelProducer` in the composable and passing all the data through Compose needlessly increases complexity and hurts performance.

> As these individual state values are loaded, the transactions are triggered, sometimes 3 or 4 in a row.

This shouldn't be the case. It points to incorrect architecture. When dealing with multiple streams, you should combine them such that you wait for the first emission from all streams before responding. See e.g. `Flow.combine`.

The `Transaction` dropping most likely relates to the frequency of the 3 or 4 updates you've mentioned and the way you run `Transaction`s. Specifically, the time between these updates is probably less than the time it takes for a `Transaction` to complete, and since you're using `LaunchedEffect` (which cancels its coroutine when the keys change), the `Transaction`s are aborted. You could rework your code so that coroutine cancelation doesn't occur, and all 3 or 4 `Transaction`s are allowed to finish. However, as mentioned above, you shouldn't have these 3 or 4 `Transaction`s in short succession in the first place — it would be preferable to address the root problem and fix the stream handling so that you run a single `Transaction` initially.

**Reply by baum-sparton** [reply]
Excellent, thanks for the feedback.

I wasn't sure of the lifetime of the extras.  Meaning I wasn't sure if old ones were dropped during a transaction or if they kept accumulating, so I was playing it safe.  I'll make those changes and move the `modelProducer` into a view model.

I can see how streamlining the compose data flow will improve performance, but I don't see how it can cause the dropped transactions.  Even as intermediate transactions are cancelled, I would expect the final one to complete and I can observe them reaching the end of the `runTransaction `block without the UI updating.  That said, I have already tested preventing runTransaction from being called until all the data is in place and in my testing the dropped transaction issue has not reoccurred.

**Reply by Gowsky** [reply]
>I would expect the final one to complete

That's correct. If the last `Transaction` was not completed, please provide an MRE, so we can look into it.

>I can observe them reaching the end of the `runTransaction` block

That doesn't mean that the `Transaction` has been completed. The lambda only specifies the data that should be added. The actual processing occurs afterward, once the lambda has returned. Therefore, what you should check is whether `runTransaction` returns.
